proposals/0698-qualified-literals.rst

```diff
@@ -0,0 +1,359 @@
+Qualified Literals
+==================
+
+.. author:: Brandon Chinn
+.. date-accepted:: Leave blank. This will be filled in when the proposal is accepted.
+.. ticket-url:: Leave blank. This will eventually be filled with the
+                ticket URL which will track the progress of the
+                implementation of the feature.
+.. implemented:: Leave blank. This will be filled in with the first GHC version which
+                 implements the described feature.
+.. highlight:: haskell
+.. header:: This proposal is `discussed at this pull request <https://github.com/ghc-proposals/ghc-proposals/pull/698>`_.
+.. sectnum::
+.. contents::
+
+This proposal proposes extending ``-XQualifiedDo`` to literals, to enable more ergonomic and more powerful syntax than ``OverloadedStrings`` or ``OverloadedLists``.
+
+Motivation
+----------
+
+``OverloadedStrings`` and ``OverloadedLists`` are useful for writing normal string/list syntax that get desugared to non-built-in types. However, they do this via typeclasses, which inherits issues like type inference ambiguity. It's possible these issues are one reason why these extensions, despite being in GHC for a long time, are not defaults in GHC202X language editions. Furthermore, they could provide even more flexibility; one note in the original ``OverloadedLists`` `design <https://gitlab.haskell.org/ghc/ghc/-/wikis/overloaded-lists>`_ mentions the inability to support heterogeneous lists.
+
+These extensions allow extensibility of string and list literals, but there's currently no option to allow extending other literals, like numeric literals. Because the ``fromInteger`` function is a part of ``Num``, there's no way to overload numeric literals without providing a specification for all the ``Num`` operations. Even if we broke out ``fromInteger`` into a separate class, we still inherit the same typeclass issues as overloaded strings/lists.
```

Comment:
I think this is another example where I think steps like this go against legibility.

`1 + 2 == 3` makes sense to me without knowing what type is involved.
`Foo.1 + Foo.2 == Foo.3`, I would expect this to behave similarly too.

In my mind, numeric syntax is defined in terms of a bunch of axioms, and that's what allows me to make sense of arbitrary code using it.

Without that I think code becomes harder to read. 

I understand wanting to split up the `Num` typeclass since not all the methods always make sense, but I think if you drop it all, you end up with something that's not really a number at all.

Reply:
You have the same problem with QualifiedDo, no?
```hs
module Foo where

(>>=) :: Int -> (User -> Bool) -> IO ()
userId >>= f = do
  u <- getUser userId
  when (f u) $ putStrLn "Found user!"

main = Foo.do
  u <- 123
  isAdmin u && isEnabled u
```
You can write do blocks that go completely against expected semantics. No one should, because it's incredibly confusing, but the power of QualifiedDo comes from its flexibility. Same argument here

Reply:
I'm curious if anyone has stories of avoiding QualifiedDo because it's possible to write weird implementations. I haven't heard any. I would think you'd just... not use weird implementations

I don't see a tension there. Being able to write `Foo.123` has no bearing on whether the resulting type supports `+`. If it does, then it should follow the laws as usual. If not, it'll be a compile error

Reply:
I see you are saying if there is `+` then it should behave as expected. That's reasonable. 

I think I hold the stronger claim that numbers should support numeric operations or at least a large subset of them, but I think that's just a point where we differ